rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+
+```nix
+some.very.long.attr = callFunction
+  arg1
+  arg2
+  arg3;
+```
+
+#### inherit
+
+**Description**
+
+The items are either all on the same line, or all on a new line each (with indentation).
+
+**Examples**
+
+```nix
+inherit foo bar baz;
+inherit
+  foo'
+  bar'
+  baz'
+;
+```
+
+#### inherit from
+
+**Description**
+
+- If the inherit target is single-line, it is placed on the same line as the `inherit`, even if the following items do not fit onto one line.
+- Otherwise, it starts on a new line with indentation, like the others.
+  - In that case, the remaining items are force-expanded too, even if they would have fit onto one line in the first place.
+
+**Examples**
+
+```nix
+inherit (pkgs) ap1 ap2 ap3;
+inherit (pkgs)
+  app1
+  app2
+  # …
+  app42
+;
+inherit
+  (pkgs.callPackage ./foo.nix {
+    arg = "val";
+  })
+  attr1
+  attr2
+;
+```
+
+#### Semicolon placement
+
+**Description**
+
+The semicolon is always placed on the same line as the expression it concludes.
+
+**Examples**
+
+```nix
+{
+  attr1 = bar;
+  attr2 = function call {
+    # stuff
+  };
+  attr3 =
+    function call
+      many
+      arguments;
+  attr4 =
+    let
+      foo = "bar";
+    in
+    some statement;
+  attr5 =
+    if foo then
+      "bar"
+    else
+      "baz";
+  attr6 =
+    let
+      foo = false;
+    in
+    if foo then "bar" else "baz";
+  attr7 = function (
+    if foo then
+      "bar"
+    else
+      "baz"
+  );
+  attr8 =
+    cond1
+    || cond2
+    ||
+      some function call
+      && cond3;
+}
+```
+
+**Rationale and alternatives**
+
+
+There are four considered semicolon styles:
+1. On a new line without indentation.
+  - This clearly marks a separation between attributes, however it is wasteful of space.
+  ```nix
+  attr3 =
+    function call
+      many
+      arguments;
+  attr3 =
+    let
+      foo = "bar";
+    in
+    some statemens;
+  ```
+2. On a new line with one indentation level.
+  - Just as wasteful on space as (1), but a bit less clear about signaling the end of the binding.
+  ```nix
+  inherit (pkgs)
+    app1
+    app2
+    # …
+    app42;
+
+  attr3 =
+    function call
+      many
+      arguments;
+  ```
+3. A mix of (1) and (2), where usually the semicolon is placed directly at the end of the binder.
+   But with exceptions in which the semicolon is placed onto the following line instead in cases where the value is a multiline `if` expression or nested operator.
+   These are the only syntax elements which may result in the semicolon being placed on a line with arbitrarily deep indentation.
+   
+   ```nix
+   attr4 =
+     if foo then
+       "bar"
+     else
+       "baz";
+
+   attr5 =
+     let
+       foo = false;
+     in
+     if foo then
+       "bar"
+     else
+       "baz";
+       
+   attr7 =
+     cond1
+     || cond2
+     ||
+       some function call
+       && cond3;
+   ```
```

Comment:
Absolutely stellar work! ❤️❤️❤️ Thanks so much for pulling through. I fully subscribe to the motivation and the full extent of the proposal. Once implemented, it will surely eliminate a significant source of friction for contributors and maintainers.

Great piece of writing, and a beautifully painted bike shed.